src/prefect/server/models/artifacts.py

```diff
@@ -48,3 +50,127 @@ async def read_artifact(
 
     result = await session.execute(query)
     return result.scalar()
+
+
+@inject_db
+async def _apply_artifact_filters(
+    query,
+    db: PrefectDBInterface,
+    flow_run_filter: filters.FlowRunFilter = None,
+    task_run_filter: filters.TaskRunFilter = None,
+    artifact_filter: filters.ArtifactFilter = None,
+):
+    """Applies filters to an artifact query as a combination of EXISTS subqueries."""
+    if artifact_filter:
+        query = query.where(artifact_filter.as_sql_filter(db))
+
+    if flow_run_filter or task_run_filter:
+        exists_clause = select(db.FlowRun).where(
+            db.Artifact.flow_run_id == db.FlowRun.id
+        )
+
+        if flow_run_filter:
+            exists_clause = exists_clause.where(flow_run_filter.as_sql_filter(db))
+        if task_run_filter:
+            exists_clause = exists_clause.join(
+                db.TaskRun,
+                db.TaskRun.flow_run_id == db.FlowRun.id,
```

Comment:
Should we be joining task run and artifact instead here?

Reply:
Ah, yes 👍